sdk/storage/azure-storage-quickquery/src/main/java/com/azure/storage/quickquery/BlobQuickQueryInputStream.java

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.quickquery;
+
+import com.azure.storage.common.ErrorReceiver;
+import com.azure.storage.common.ProgressReceiver;
+import com.azure.storage.quickquery.models.BlobQuickQueryError;
+import org.apache.avro.file.DataFileStream;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+
+public class BlobQuickQueryInputStream extends InputStream {
+
+    private DataFileStream<GenericRecord> parsedStream;
+    private ByteArrayInputStream userBuffer;
+    private boolean endRecordSeen;
+    private boolean firstRead;
+    private InputStream networkStream;
+
+    private final ErrorReceiver<BlobQuickQueryError> errorReceiver;
+    private final ProgressReceiver progressReceiver;
+
+    /**
+     * Creates a new {@code BlobQuickQueryInputStream}
+     * @param networkStream {@link InputStream network stream}
+     * @param errorReceiver {@link ErrorReceiver<BlobQuickQueryError>}
+     * @param progressReceiver {@link ProgressReceiver}
+     */
+    BlobQuickQueryInputStream(InputStream networkStream, ErrorReceiver<BlobQuickQueryError> errorReceiver,
+        ProgressReceiver progressReceiver) {
+        this.networkStream = networkStream;
+        this.errorReceiver = errorReceiver;
+        this.progressReceiver = progressReceiver;
+        this.endRecordSeen = false;
+        this.firstRead = true;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        validateParameters(b, off, len);
+
+        /* If len is 0, then no bytes are read and 0 is returned. */
+        if (len == 0) {
+            return 0;
+        }
+        /* Attempt to read at least one byte. If no byte is available because the stream is at end of file,
+           the value -1 is returned; otherwise, at least one byte is read and stored into b. */
+
+        /* First read? */
+        if (firstRead) {
+            firstRead = false;
+            /* Reads from the network to determine schema. */
+            parsedStream = new DataFileStream<>(networkStream, new GenericDatumReader<>());
+            this.userBuffer = new ByteArrayInputStream(new byte[0]);
+        }
+
+        /* Now we are guaranteed that buffer is SOMETHING. */
+        /* No data is available in the buffer.  */
+        if (this.userBuffer.available() == 0) {
+            /* Try to get more data */
```

Comment:
I think adding a comment that we need the while loop because not every record gives us data, and we need to keep parsing until we find a data record would be helpful